### DISCUSSION TITLE
[QoL] Automatically reduce project download sizes

### DISCUSSION THREAD

**Discussion by MIfoodie** [godotengine/godot-proposals#7114] [Engine Core]
I was reading the [manual](https://docs.godotengine.org/en/stable/contributing/development/compiling/optimizing_for_size.html#doc-optimizing-for-size) where I found that you can disable the 3D engine and disable advanced GUI objects to reduce the size of you project. My idea is for Godot to do this automatically.

When the user exports the project (in any format), Godot should check for any object, node, script, etc in the 3D space. If there is none, it will remove the 3D engine from the download to reduce the size of the download. The same could be done with advanced GUI objects, where if the project doesn’t use them, they could automatically be disabled.

**Comment by theraot** [comment]
Something like this has come up before. The closer I could find is #1001 

I believe right now this feature requires more export templates (variations with and without 3D).

<details><summary>Details</summary>
<p>

Also, export templates for every combination of features we want to make this with. Which at some point means that it is better to have Godot download its source code an automate the template build process. *In fact, the day that C# does not require a separate build might come. But not this evening.*

</p>
</details>

If this is done for only one feature, 3D is likely the best option.

By the way, without the checks, just offering the developer to export without 3D would go a long way (the developer likely knows if they are using 3D, and may also test with and without 3D to check if it causes problems). Even if Godot checks, I would want the developer to be able to decide.

So the work flow I would want would be:

1. Go to the export settings,
2. There is UI where you can disable 3D, and the option to let Godot check.
3. If the developer lets Godot check, it will either automatically disable 3D or not.
4. Otherwise we can do it manually.

**Reply by Calinou** [reply]
[Feature build profiles](https://github.com/godotengine/godot/pull/62996) already exist and can be generated by the editor, but the buildsystem needs further changes before it can really be effective. Also, no matter what, it will still require you to recompile export templates for all platforms you wish to export to.

**Reply by theraot** [reply]
@Calinou That is what I was looking for but could not find, thank you.

**Comment by Zireael07** [comment]
I asked for something similar for WASM export and was told this requires recompiling export templates and providing them to end-users yourself. So... not quite automatic.
